materialize-mongodb/transactor.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"encoding/json"
+	"context"
+	"os"
+	"fmt"
+	"math"
+
+	pm "github.com/estuary/flow/go/protocols/materialize"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type transactor struct {
+	bindings []*binding
+}
+
+type binding struct {
+	collection *mongo.Collection
+}
+
+var idField = "_id"
+
+func (t *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	for it.Next() {
+		var key = string(it.Key.Pack())
+		var collection = t.bindings[it.Binding].collection
+
+		var ctx = context.Background()
+		var cur, err = collection.Find(ctx, bson.D{{idField, bson.D{{"$eq", key}}}})
+		defer cur.Close(ctx)
+		if err != nil {
+			return fmt.Errorf("finding document in collection: %w", err)
+		}
+		if !cur.Next(ctx) {
+			continue
+		}
+
+		var doc bson.M
+
+		if err = cur.Decode(&doc); err != nil {
+			return fmt.Errorf("decoding document in collection %s: %w", collection.Name(), err)
+		}
+		doc = sanitizeDocument(doc)
+
+		js, err := json.Marshal(doc)
+		if err != nil {
+			return fmt.Errorf("encoding document in collection %s as json: %w", collection.Name(), err)
+		}
+
+		if err := loaded(it.Binding, js); err != nil {
+			return fmt.Errorf("sending loaded: %w", err)
+		}
+	}
+
+	return it.Err()
+}
+
+func (t *transactor) Store(it *pm.StoreIterator) (pm.StartCommitFunc, error) {
```

Comment:
It looks to me like we'll need to do something in the transactor to allow for exactly-once semantics with these loads & updates, otherwise zombie shards would end up racing to re-apply/overwrite & undo updates. I don't know mongodb in-depth, but I wonder if it would support enough transactional ability to serve as an [authoritative remote store](https://docs.estuary.dev/reference/Connectors/materialization-protocol/#common-implementation-patterns)?

Reply:
@williamhbaker I had not written such a materialization before (the ones I had worked on were recovery log + idempotent apply), so I tried to implement the fencing mechanism for MongoDB. Let me know if it makes sense and the implementation looks okay to you.

There is no `KeyBegin` and `KeyEnd` handling since we don't need to run multiple instances of this connector (at least not now). We can revisit that later if necessary.

Reply:
Made a couple more comments, looks like it is pretty close. Ignoring materialization sharding for now seems reasonable. I guess the fencing mechanism sans key ranges would cause split shards to fail in that case, but at least they wouldn't overwrite each other's data.